app/slashcommands/util.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package slashcommands
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+const (
+	ActionKey = "-action"
+)
+
+// responsef creates an ephemeral command response using printf syntax.
+func responsef(format string, args ...interface{}) *model.CommandResponse {
+	return &model.CommandResponse{
+		ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL,
+		Text:         fmt.Sprintf(format, args...),
+		Type:         model.POST_DEFAULT,
+	}
+}
+
+// parseNamedArgs parses a command string into a map of arguments. It is assumed the
+// command string is of the form `<action> --arg1 value1 ...` Supports empty values.
+// Arg names are limited to [0-9a-zA-Z_].
+func parseNamedArgs(cmd string) map[string]string {
+	m := make(map[string]string)
+
+	split := strings.Fields(cmd)
+
+	// check for optional action
+	if len(split) >= 2 && !strings.HasPrefix(split[1], "--") {
+		m[ActionKey] = split[1] // prefix with hyphen to avoid collision with arg named "action"
+	}
+
+	// To support values containing hyphens, we allow all characters up until the next double hyphen.
+	// However Go regex does not support negative lookahead, so we fake it by adding extra double hyphens
+	// which get eaten by the non-capturing group.
+	cmdFixed := strings.ReplaceAll(cmd, "--", "-- --") + " --"
+
+	re := regexp.MustCompile(`(\-\-\w+\s+)(.*?)(?:\-\-)`) //re := regexp.MustCompile(`(\-\-\w+\s+)([^\-]*)`)
+	splitArgs := re.FindAllStringSubmatch(cmdFixed, -1)
+
+	for _, arr := range splitArgs {
+		arg := trimSpaceAndQuotes(arr[1][2:]) // strip out the leading hyphens
+		val := trimSpaceAndQuotes(arr[2])
+
+		if arg != "" {
+			m[arg] = val
+		}
+	}
+	return m
```

Comment:
I generally like regexps but I am wondering if a more iterative approach could be easier to digest than the _black magic_.

Something like:

```go
	for i := 0; i < len(split); i++ {
		if !strings.HasPrefix(split[i], "--") {
			continue
		}
		var val string
		arg := trimSpaceAndQuotes(strings.Trim(split[i], "-"))
		if i < len(split)-1 && !strings.HasPrefix(split[i+1], "--") {
			val = trimSpaceAndQuotes(split[i+1])
		}
		if arg != "" {
			m[arg] = val
		}
	}
```

Reply:
TBH, I read and grok the regex much more easily than the for loop with special case conditionals and off-by-one issues.  Your version works, but I had to mentally loop through that multiple times to really make sure.

That said, I understand a lot of people don't read regex anymore so I'll change it to your pure-Go way.  Thanks.

Reply:
Agreed. The lack of negative look-ahead forced a bit of a hack that took away the elegance of a clean regex. Ah well.